lib/galaxy/objectstore/_caching_base.py

```diff
@@ -0,0 +1,384 @@
+
+        # for JOB_WORK directory
+        if obj_dir:
+            rel_path = os.path.join(rel_path, str(object_id))
+        if base_dir:
+            base = self.extra_dirs.get(base_dir)
+            assert base
+            return os.path.join(base, rel_path)
+
+        if not dir_only:
+            rel_path = os.path.join(rel_path, alt_name if alt_name else f"dataset_{object_id}.dat")
+
+        if in_cache:
+            return self._get_cache_path(rel_path)
+
+        return rel_path
+
+    def _get_cache_path(self, rel_path: str) -> str:
+        return os.path.abspath(os.path.join(self.staging_path, rel_path))
+
+    def _in_cache(self, rel_path: str) -> bool:
+        """Check if the given dataset is in the local cache and return True if so."""
+        cache_path = self._get_cache_path(rel_path)
+        return os.path.exists(cache_path)
+
+    def _pull_into_cache(self, rel_path) -> bool:
+        ipt_timer = ExecutionTimer()
+        # Ensure the cache directory structure exists (e.g., dataset_#_files/)
+        rel_path_dir = os.path.dirname(rel_path)
+        if not os.path.exists(self._get_cache_path(rel_path_dir)):
+            os.makedirs(self._get_cache_path(rel_path_dir), exist_ok=True)
+        # Now pull in the file
+        file_ok = self._download(rel_path)
+        fix_permissions(self.config, self._get_cache_path(rel_path_dir))
+        log.debug("_pull_into_cache: %s\n\n\n\n\n\n", ipt_timer)
+        return file_ok
+
+    def _get_data(self, obj, start=0, count=-1, **kwargs):
+        rel_path = self._construct_path(obj, **kwargs)
+        # Check cache first and get file if not there
+        if not self._in_cache(rel_path):
+            self._pull_into_cache(rel_path)
+        # Read the file content from cache
+        data_file = open(self._get_cache_path(rel_path))
+        data_file.seek(start)
+        content = data_file.read(count)
+        data_file.close()
+        return content
+
+    def _exists(self, obj, **kwargs):
+        in_cache = exists_remotely = False
+        rel_path = self._construct_path(obj, **kwargs)
+        dir_only = kwargs.get("dir_only", False)
+        base_dir = kwargs.get("base_dir", None)
+
+        # check job work directory stuff early to skip API hits.
+        if dir_only and base_dir:
+            if not os.path.exists(rel_path):
+                os.makedirs(rel_path, exist_ok=True)
+            return True
+
+        in_cache = self._in_cache(rel_path)
+        exists_remotely = self._exists_remotely(rel_path)
+        dir_only = kwargs.get("dir_only", False)
+        base_dir = kwargs.get("base_dir", None)
+        if dir_only:
+            if in_cache or exists_remotely:
+                return True
+            else:
+                return False
+
+        # TODO: Sync should probably not be done here. Add this to an async upload stack?
+        if in_cache and not exists_remotely:
+            self._push_to_os(rel_path, source_file=self._get_cache_path(rel_path))
```

Comment:
Is it time to also rename `_push_to_os` to maybe `_push_to_storage` or `_push_to_remote` ? I always get confused by that